bitcoin/src/address/mod.rs

```diff
@@ -413,55 +436,57 @@ impl Address {
         if script.len() > MAX_SCRIPT_ELEMENT_SIZE {
             return Err(Error::ExcessiveScriptSize);
         }
-        let payload = Payload::ScriptHash(script.script_hash());
-        Ok(Address(AddressInner { network, payload }, PhantomData))
+        let hash = script.script_hash();
+        Ok(Address::_p2sh(hash, network))
+    }
+
+    fn _p2sh(hash: ScriptHash, network: Network) -> Address {
+        let prefix = LegacyP2shPrefix::from_network(network);
+        Self(AddressInner::P2sh { hash, prefix }, PhantomData)
     }
 
     /// Creates a witness pay to public key address from a public key.
     ///
     /// This is the native segwit address type for an output redeemable with a single signature.
     ///
     /// # Errors
+    ///
     /// Will only return an error if an uncompressed public key is provided.
     pub fn p2wpkh(pk: &PublicKey, network: Network) -> Result<Address, Error> {
-        let prog = WitnessProgram::new(
-            WitnessVersion::V0,
-            pk.wpubkey_hash().ok_or(Error::UncompressedPubkey)?,
-        )?;
-        let payload = Payload::WitnessProgram(prog);
-        Ok(Address(AddressInner { network, payload }, PhantomData))
+        let hash = pk.wpubkey_hash().ok_or(Error::UncompressedPubkey)?;
+        let program = WitnessProgram::new(WitnessVersion::V0, hash)
+            .expect("wpubkey hash is a valid 20 byte program for segwit v0");
```

Comment:
For consideration: add a constructor asserting this to the `WitnessProgram` so it can be used by other crates. Then the checking logic is at one place. (And we could even skip it.)

Other than that, this is indeed nicer.

Reply:
Like `WitnessProgram::new_unchecked` you mean? And put the contents of the `expect` above it in a code comment?

Reply:
`new_v0` and `new_v1` without the version arg would be nice too.

Reply:
No, I meant to have `WitnessProgram::p2wpkh` which lives near `WitnessProgram` and then just call that one here.

Reply:
This turned out quite nicely, thanks for the suggestion!